### PR TITLE
WarpOS support

### PR DESCRIPTION
**Pull request by MBeijer**
@ajenbo @arczi84 Here comes support for WarpOS as well, oh boy

### CODE REVIEW

File: CMakeLists.txt
Comment (MBeijer) [inline CMakeLists.txt]: Please fix the newline on this, as now it looks like the whole file has changed

File: CMakeLists.txt
Comment (MBeijer) [inline CMakeLists.txt]: This might not be needed as it checks if LTO is available below

File: CMakeLists.txt
Comment (MBeijer) [inline CMakeLists.txt]: Please fix this
```suggestion
    if(NOT WARPOS)
```

File: CMake/amiga/ppc-warpos.cmake
Comment (MBeijer) [inline CMake/amiga/ppc-warpos.cmake]: In order for this to compile on other systems that MorphOS, may I suggest adding the WarpOS prefix path as an option?
```suggestion
option(WARPOS_PREFIX "Path to WarpOS prefix" "/gg")
set(CMAKE_SYSROOT "${WARPOS_PREFIX}/ppc-warpos")
set(CMAKE_FIND_ROOT_PATH "${WARPOS_PREFIX}/ppc-warpos")
set(tools "${WARPOS_PREFIX}/bin/")
```

File: CMake/amiga/ppc-warpos.cmake
Comment (MBeijer) [inline CMake/amiga/ppc-warpos.cmake]: on cross tools this is ppc-warpos-gcc perhaps an option to add this as well?

File: CMakeLists.txt
Comment (Cowcat5150) [reply CMakeLists.txt]: Cmake doesn't find it: Unknown CMake command "check_ipo_supported"

File: CMakeLists.txt
Comment (Cowcat5150) [reply CMakeLists.txt]: oops...

**Comment by AJenbo** [comment]
Not to be confused with OS/2 Warp I take it :)

File: SourceX/DiabloUI/selhero.cpp
Comment (AJenbo) [inline SourceX/DiabloUI/selhero.cpp]: Let's not merge this before it compiles :)

File: SourceX/dx.cpp
Comment (AJenbo) [inline SourceX/dx.cpp]: Is there something like:
```suggestion
#if defined (__AMIGA__) && defined(__M68K__)
```
That would probably be more accurate?

File: readme.warpos
Comment (AJenbo) [inline readme.warpos]: Probably this file should live in Packaging/WarpOS or Packaging/Amiga

**Review by AJenbo — COMMENTED** [review]
Could you make a PR against Diasurgical/DevilutionX with the missing readme/license files, that way this PR should become a lot smaller.

Also IMHO it would be better to finalize the main port before stacking more derivatives on top, so maybe best to keep this as a draft for now :)

File: CMake/amiga/ppc-warpos.cmake
Comment (Cowcat5150) [reply CMake/amiga/ppc-warpos.cmake]: Nowadays only newer gcc/g++ for warpos exist through mos2wos (newlib). But no harm with suggestion if this option works in cmake MOS.

File: SourceX/DiabloUI/selhero.cpp
Comment (Cowcat5150) [reply SourceX/DiabloUI/selhero.cpp]: Problems with chronos include. A simple test for chronos in MOS works but not for warpos SDK.

File: SourceX/dx.cpp
Comment (Cowcat5150) [reply SourceX/dx.cpp]: Could be, but technically no PPC system should see this: In rare cases (MOS) 68k code could be used when compiling emulation code. Don't know if that could trigger some hidden includes in MOS SDK.

File: CMakeLists.txt
Comment (Cowcat5150) [reply CMakeLists.txt]: I will reuse the original file then. All source is "suffering" from "dos cr": Apart from that, using some editors the whole source has big spacing between line codes. Maybe I will use another editor.

Comment (MBeijer) [reply CMakeLists.txt]: the line endings is converted by git on Windows if you don't tell it to ignore it. It's a pain in the ass. I've configured my computers to always keep line endings as is when checking out to avoid this.

Comment (Cowcat5150) [reply CMakeLists.txt]: Old school CygnusEd and others see all that but not the hidden characters: Even some of my changes were marked in "red" by github. Not good.

Comment (AJenbo) [reply CMakeLists.txt]: It using CR since we expect that to have been the case for the original, it's-not-a-bug-it's-a-feature :)

**Review by arczi84 — DISMISSED** [review]

File: readme.warpos
Comment (Cowcat5150) [reply readme.warpos]: Probably but that's why I did a fork: To not mess with the original before a proper final port is done, meanwhile people can try it and give some input. My changes don't use 8bit screen mode as this doesn't work in my environment (wrong colors and unable to do fullscreen). I even did a unique Makefile for the whole project instead of cmake.........Just take what could be relevant meanwhile I would try to do the moving target that is a MOS binary with cmake.

Comment (AJenbo) [reply readme.warpos]: Colors are fixed by this commit https://github.com/diasurgical/devilutionX/commit/319bdeddd242bb79c5145fd81ee68d584d4b220e

File: SourceX/miniwin/misc.cpp
Comment (AJenbo) [inline SourceX/miniwin/misc.cpp]: I'll try and upstream this as it appears to affect multiple SDL 1.2 targets

Comment (Cowcat5150) [reply SourceX/miniwin/misc.cpp]: An issue in warpos sdl library : It's unclear if it could affect real Amigas in some situations, but certainly MOS machines running warpos programs, because game stops right there as other test examples showed.

Comment (AJenbo) [reply SourceX/miniwin/misc.cpp]: It's also affecting Nintendo 3DS

File: readme.warpos
Comment (Cowcat5150) [reply readme.warpos]: You're correct but there's a problem: In window mode, besides not so good colors when you have a 16 bit desktop, movie starts with his own resolution not (scaled for) 640x480 as menu has when appears. But when you select fullscreen in diablo.ini, game crashes possibly due to the inability in my system to choose that small movie screen resolution. In non 8 bit mode, movies are scaled to the proper game resolution in window and fullscreen.

File: SourceX/miniwin/misc.cpp
Comment (Cowcat5150) [reply SourceX/miniwin/misc.cpp]: Why the need of cdrom then? :)

Comment (MBeijer) [reply SourceX/miniwin/misc.cpp]: @Cowcat5150 that's why it's going to be upstreamed for all SDL1.2 targets. :P CD is not really needed for devilutionX no.

File: CMakeLists.txt
Comment (MBeijer) [inline CMakeLists.txt]: @AJenbo are you fine with this? perhaps 3.10 is required for the LTO check? Otherwise it seems to work on 3.8

File: CMakeLists.txt
Comment (MBeijer) [inline CMakeLists.txt]: Perhaps LTO check should be checked below all the different platforms so we can disable LTO on platforms that it's known not to work on?

File: SourceX/dx.cpp
Comment (MBeijer) [reply SourceX/dx.cpp]: @Cowcat5150 if you compile for PPC, __M68K__ should never be set. The MorphOS GCC is 100% PPC only, so I don't know why it would enable 68k emulation.

File: readme.warpos
Comment (MBeijer) [reply readme.warpos]: @Cowcat5150 when executing SDL_SetVideoMode, if the requested screenmode doesn't exist, it should default to the lowest possible/available, it shouldn't crash. I would issue this as a WarpOS SDL bug.

**Comment by MBeijer** [comment]
@Cowcat5150 Please remove the devilutionx-wos.lha binary package from the repo and add it as a release here instead: https://github.com/Cowcat5150/devilutionX/releases

File: SourceX/miniwin/misc.cpp
Comment (AJenbo) [reply SourceX/miniwin/misc.cpp]: SDL_INIT_CDROM isn't even a thing in SDL2 (the main target)

File: CMakeLists.txt
Comment (AJenbo) [reply CMakeLists.txt]: Ubuntu 18.04, Android Stuido, Visual Stuido and possibly Xcode are all stuck on 3.10 so that is our upper limit. CheckIPOSupported appears to have been introduced in 3.9 so I guess that is our lower limit. This also explains why @Cowcat5150 was having issues with LTO detection if he is running CMake 3.8. @Cowcat5150 is it possibl for you to upgrade to at least 3.9?

Comment (MBeijer) [reply CMakeLists.txt]: MorphOS SDK seems to be stuck at 3.8. On crosstools this shouldn't be an issue though. But my suggestion to put Platform defines above LTO check could be a workaround on systems where this is not supported. (as stated in another comment below)

File: SourceX/dx.cpp
Comment (Cowcat5150) [reply SourceX/dx.cpp]: @MBeijer I know, I know. It's the "but if" in me. Just to clarify to others that when "__AMIGA__" is defined it implies MOS too. By the same token if "__MORPHOS__" is defined it implies Warpos in "our" particular SDK (you can differentiate it with WARPUP) so flags are going back and forth for what kind of system are you compiling. Not an easy planet.

Comment (Cowcat5150) [reply SourceX/dx.cpp]: Shit the underscores don't show here...

File: readme.warpos
Comment (Cowcat5150) [reply readme.warpos]: But the movie playing in 8bit mode should be scaled to 640x480? Crash or not is another kind of animal when warpos is running in MOS. I don't know at this moment.

File: CMakeLists.txt
Comment (Cowcat5150) [reply CMakeLists.txt]: I cannot upgrade to upper cmake till a newer warpos SDK is done against newer MOS SDK that has 3.10 cmake.

File: readme.warpos
Comment (MBeijer) [reply readme.warpos]: No, the resolution changes to 320x200 because it takes less resources from a low end system than upscaling to 640x480. 8bit mode in devilitionX is “low end system mode”. And 8bit screen also takes less memory bandwidth. There’s nothing in the game as is that is more than 256 colors either.

File: CMakeLists.txt
Comment (MBeijer) [reply CMakeLists.txt]: Download the mos SDK and compile cmake from there yourself? Shouldn’t be too hard.

Comment (Cowcat5150) [reply CMakeLists.txt]: I could try to just copy newer cmake package and see what happens, but installing newer SDK over the old and being ok with it thinking that nothing is gonna happen with the warpos side is like making christmas cakes with cocaine instead of sugar.

Comment (Cowcat5150) [reply CMakeLists.txt]: And what the hell is "dismissed review" here? I don't dismiss anyone. :)

Comment (Cowcat5150) [reply CMakeLists.txt]: Crap, cmake from newer MOS still 3.8.....

File: readme.warpos
Comment (Cowcat5150) [reply readme.warpos]: Morphos by default has the lower screen limit set to 640x480 for all bit depths. I will test what happens when you have Ambient set to only 8bit without having game hardcoded to 8bit or if there's a way to make a Radeon promote a 320x200 mode.

Comment (MBeijer) [reply readme.warpos]: If 640x480 is lowest, that resolution is picked. However there’s no scaling so the video will be centered and small

Comment (Cowcat5150) [reply readme.warpos]: Should be. It's a matter to test that when a MOS port will exist.

File: CMakeLists.txt
Comment (Cowcat5150) [reply CMakeLists.txt]: Setting if(NOT WARPOS) on top Cmakelists for choosing 3.10 or 3.8 cmake_minimum_ is useless.

File: CMakeLists.txt
Comment (Cowcat5150) [reply CMakeLists.txt]: Here if(NOT WARPOS) works.

**Comment by MBeijer** [comment]
@Cowcat5150 when doing updates, do a merge or rebase rather than copy the changed files manually. It can cause conflicts and the commit history gets messy.

**Comment by Cowcat5150** [comment]
@MBeijer For unknown reasons, files in some dirs got copied without capital letters and as a result there were 2 files for each one. This is the strange last commit.

**Comment by Cowcat5150** [comment]
@MBeijer @AJenbo Latest commit (of yours) comes with some problems: itoa seems to be missing when trying to compile (Source/control.cpp SDL_itoa call). Reenabling char *itoa like it was in old miniwin/misc.h gets through to the link stage, but then no itoa function is found. Had to recreate an itoa function....  Then, calling cmake .. -DBINARY_RELEASE as I do, doesn't work as it acts like debug on, faster on, dist off, etc.....And dreaded gnu++_concurrence_error starts to show for MOS emulating Warpos (a known bug).

**Comment by MBeijer** [comment]
> @MBeijer @AJenbo Latest commit (of yours) comes with some problems: itoa seems to be missing when trying to compile (Source/control.cpp SDL_itoa call). Reenabling char *itoa like it was in old miniwin/misc.h gets through to the link stage, but then no itoa function is found. Had to recreate an itoa function.... Then, calling cmake .. -DBINARY_RELEASE as I do, doesn't work as it acts like debug on, faster on, dist off, etc.....And dreaded gnu++_concurrence_error starts to show for MOS emulating Warpos (a known bug).

SDL_itoa has been replaced with snprintf now

**Comment by Cowcat5150** [comment]
> > @MBeijer @AJenbo Latest commit (of yours) comes with some problems: itoa seems to be missing when trying to compile (Source/control.cpp SDL_itoa call). Reenabling char *itoa like it was in old miniwin/misc.h gets through to the link stage, but then no itoa function is found. Had to recreate an itoa function.... Then, calling cmake .. -DBINARY_RELEASE as I do, doesn't work as it acts like debug on, faster on, dist off, etc.....And dreaded gnu++_concurrence_error starts to show for MOS emulating Warpos (a known bug).
> 
> SDL_itoa has been replaced with snprintf now

@MBeijer @AJenbo This is what I got that make stuff crash, specially drlg_l4 in higher levels:

/Source/control.cpp:1728:33: runtime error: shift exponent 33 is too large for 32-bit type 'int'


/Source/drlg_l4.cpp:1236:21: runtime error: index 20 out of bounds for type 'unsigned char [20]'

/Source/missiles.cpp:3118:39: runtime error: left shift of negative value -32
/Source/missiles.cpp:3119:39: runtime error: left shift of negative value -16

control is more about first time you ask magic menu. The others happen I think towards level 12+. I missed to catch some runtime error in drlg_l1 too....

Now I see control.cpp runtime error happens when in god/cheat mode (full magic, etc)....

**Comment by MBeijer** [comment]
You're clearly not making things easy for yourself when merging in new changes... It's hard for me to see what potentially went missing during your manual merge, that could potentially cause this. @AJenbo made some changes for this to work on Big Endian PPC (linux) so it should work on WarpOS as well.

**Comment by MBeijer** [comment]
If you give me read/write access to your fork, I can help you with that.

**Comment by AJenbo** [comment]
/Source/control.cpp:1728 was fixed in Hellfire and we will likely apply the fix soonish

**Comment by Cowcat5150** [comment]
> If you give me read/write access to your fork, I can help you with that.

My fork is frozen because there are too many changes and conflicts and everytime I have to redoo all again. Forget about this for now. The point I'm making is everytime I play the higher levels, game crashes sooner or later: Now I'm testing it from my debian build and those runtime errors happens "unharmed" for x86 but those are (or one of) the culprit of crashes on MOS. From Level 13, calls to drlg_l4.cpp start the problems. And all is done with new github release with (minimal now) changes for it to run on MOS.

**Comment by MBeijer** [comment]
With the appropriate tool, you can merge when there are conflicts. I know
how to do this.

**Comment by Cowcat5150** [comment]
Upsss....I touched something.

**Comment by Cowcat5150** [comment]
Well, I checked those "runtime errors". Who fixes then the problems?

**Comment by AJenbo** [comment]
Anyone who finds a proper solution

**Comment by Cowcat5150** [comment]
> Anyone who finds a proper solution

I have a dumb one: In L4drawRoom function, dung[i + x][j + y] = 1 gets 20 for (j + y) so it's OOB so I force it to read 19 when this happens. Don't know if it is a solution.

Now the next problem I see that gets downhill is single_*.sv starts to decrement in size in one particular moment till further down the line you cannot read it anymore in MOS. It does in debian too, so I don't know what it is, because not a trace of problem. I notice it because MOS gets tmp files in .recicled/ and saw the whole bunch of files that got deleted when playing (I copied to debian the last deleted file with higher size).

and plus another runtime: /Source/monster.cpp:2129:30: runtime error: left shift of negative value -4


I have an idea of what is happening: tmp files in between aren't created. Long story short, occasionally ShowOutOfDiskError() messages happen in a random way.

**Pull request closed without merging**
